Zend/zend_call_stack.c

```diff
@@ -0,0 +1,426 @@
+		max_size = prev_end - end;
+	}
+
+	stack->base = (void*)end;
+	stack->max_size = max_size;
+
+	return true;
+}
+
+static bool zend_call_stack_get_linux(zend_call_stack *stack)
+{
+	if (zend_call_stack_is_main_thread()) {
+		return zend_call_stack_get_linux_proc_maps(stack);
+	}
+
+	return zend_call_stack_get_linux_pthread(stack);
+}
+#else /* __linux__ */
+static bool zend_call_stack_get_linux(zend_call_stack *stack)
+{
+	return false;
+}
+#endif /* __linux__ */
+
+#ifdef __FreeBSD__
+static bool zend_call_stack_is_main_thread(void)
+{
+	int is_main = pthread_main_np();
+	return is_main == -1 || is_main == 1;
+}
+
+# if defined(HAVE_PTHREAD_ATTR_GET_NP) && defined(HAVE_PTHREAD_ATTR_GET_STACK)
+static bool zend_call_stack_get_freebsd_pthread(zend_call_stack *stack)
+{
+	pthread_attr_t attr;
+	int error;
+	void *addr;
+	size_t max_size;
+	size_t guard_size;
+
+	/* pthread will return bogus values for the main thread */
+	ZEND_ASSERT(!zend_call_stack_is_main_thread());
+
+	pthread_attr_init(&attr);
+
+	error = pthread_attr_get_np(pthread_self(), &attr);
+	if (error) {
+		goto fail;
+	}
+
+	error = pthread_attr_getstack(&attr, &addr, &max_size);
+	if (error) {
+		goto fail;
+	}
+
+	stack->base = (int8_t*)addr + max_size;
+	stack->max_size = max_size;
+
+	pthread_attr_destroy(&attr);
+	return true;
+
+fail:
+	pthread_attr_destroy(&attr);
+	return false;
+}
+# else /* defined(HAVE_PTHREAD_ATTR_GET_NP) && defined(HAVE_PTHREAD_ATTR_GET_STACK) */
+static bool zend_call_stack_get_freebsd_pthread(zend_call_stack *stack)
+{
+	return false;
+}
+# endif /* defined(HAVE_PTHREAD_ATTR_GET_NP) && defined(HAVE_PTHREAD_ATTR_GET_STACK) */
+
+static bool zend_call_stack_get_freebsd_sysctl(zend_call_stack *stack)
+{
+	void *stack_base;
+	int mib[2] = {CTL_KERN, KERN_USRSTACK};
+	size_t len = sizeof(stack_base);
+	struct rlimit rlim;
+
+	/* This method is relevant only for the main thread */
+	ZEND_ASSERT(zend_call_stack_is_main_thread());
+
+	if (sysctl(mib, sizeof(mib)/sizeof(*mib), &stack_base, &len, NULL, 0) != 0) {
+		return false;
+	}
+
+	if (getrlimit(RLIMIT_STACK, &rlim) != 0) {
+		return false;
+	}
+
+	if (rlim.rlim_cur == RLIM_INFINITY) {
+		return false;
+	}
+
+	size_t guard_size = getpagesize();
+
+	stack->base = stack_base;
+	stack->max_size = rlim.rlim_cur - guard_size;
+
+	return true;
+}
+
+static bool zend_call_stack_get_freebsd(zend_call_stack *stack)
+{
+	if (zend_call_stack_is_main_thread()) {
+		return zend_call_stack_get_freebsd_sysctl(stack);
+	}
+
+	return zend_call_stack_get_freebsd_pthread(stack);
+}
+#else
+static bool zend_call_stack_get_freebsd(zend_call_stack *stack)
+{
+	return false;
+}
+#endif /* __FreeBSD__ */
+
+#ifdef ZEND_WIN32
+static bool zend_call_stack_get_win32(zend_call_stack *stack)
+{
+	MEMORY_BASIC_INFORMATION stack_info;
+	int8_t *base;
+
+#ifdef _M_ARM64
+	return false;
+#endif
+
+#ifdef _M_X64
+	base = (void*)((NT_TIB64*)NtCurrentTeb())->StackBase;
+#else
+	base = (void*)((NT_TIB*)NtCurrentTeb())->StackBase;
```

Comment:
I'm uncomfortable to use any `Nt*` functions; can't we use [`GetCurrentThreadStackLimits()`](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getcurrentthreadstacklimits) instead?

Reply:
> Are the requirements ok in your opinion? (_WIN32_WINNT >= 0x0602, Windows 8 || Windows Server 2012)

Oh, I haven't noticed that, but thinking about it should be okay to drop support for Windows 7 and Server 2008 shouldn't be an issue; these are no longer supported by Microsoft as of 2020-01-14 anyway.



> I think that this would bring ARM64 support as well, although I will not be able to test.

Yeah, that is a general issue though. Just assume it works, and if anybody reports issues, we can still try to address these.